gef.py

```diff
@@ -7910,44 +7910,39 @@ def __init__(self):
 
     @only_if_gdb_running
     def do_invoke(self, argv):
-        argc = len(argv)
-        if argc < 2:
-            self.usage()
-            return
-
-        arg0, argv = argv[0].lower(), argv[1:]
+        fmt = "byte"
+        target = "$sp"
         valid_formats = ["byte", "word", "dword", "qword"]
-        fmt = None
-        for valid_format in valid_formats:
-            if valid_format.startswith(arg0):
-                fmt = valid_format
-                break
-        if not fmt:
-            self.usage()
-            return
+        read_len = None
+        reverse = False
+        
+        for arg in argv:
+            arg = arg.lower()
+            is_format_given = False
+            for valid_format in valid_formats:
+                if valid_format.startswith(arg):
+                    fmt = valid_format
+                    is_format_given = True
+                    break
+            if is_format_given:
```

Comment:
why don't you keep the `if not fmt` test, but just continue instead of returning?

Reply:
The fmt variable is not None anymore by default. It is set to byte as default. Therefore, there is no need to print usage if the format is not given. If it is not given, then it is assumed that it is byte.

Reply:
Ok, I see.

But won't this allow you to provide more than one format? (only using the last one) I think `is_format_given` should be initialized outside the argv loop, and the inner valid_formats loop should be skipped if it's set. This'll avoid checking every format against every arg.

Reply:
It is the inteded behaviour. If we want to specify arguments and make them optional and position-independent at the same time, we need to use a parsers and we need some sort of indicators for example, `hexdump -f byte -n 12 -a 0xdeadbeef` which is alike linux command line tools. However, it makes it a bit harder to use. Without that kind of specification, it is better to accept the last valid parameter as accepted. If someone types `hexdump byte dword qword byte $rbp $rsp $rax`, It is convenient to parse it as `hexdump byte $rax`, instead of raising an error. Since the first priority is being easy to use, we need to assume that the user knows what he is doing. After learning the tool, this way will be easier and faster rather than repeating an order or putting extra stuff like -f, -a, -n etc.

Also, if initialize `is_format_given` outside the argv loop, then the parser will assume the given extra format is a target. For example, `hexdump dword $rbp qword` will assume that `qword` is the target. You may ask why we do think about such a scenerio which obviously looks like a misusage of the arguments. Well, let's say I just typed a long command such as `hexdump word $rbp+0x20 10`. Then, I decided to parse it as dword instead. Instead of typing everything again, I can just use the up arrow key to bring the previous command and add the new parameters at the end of it such as `hexdump word $rbp+0x20 10 qword`. The same idea applies to all parameters.